_episodes/38.md

```diff
@@ -0,0 +1,475 @@
+  time:
+---
+
+## Guests
+
+In this episode we have the pleasure to chat with a couple familiar faces who
+have been hard at work building and understanding the features we're talking
+about today:
+
+### [Kasia Findeisen](https://github.com/kasiafi), Trino Maintainer
+
+[TODO - Kasia]:<>
+
+### [Martin Traverso](https://twitter.com/mtraverso), Trino Co-Creator & Maintainer
+
+[TODO - Martin]:<>
+
+## Releases 387 to 391
+
+Official highlights from Martin:
+
+[TODO - Martin]:<>
+
+More detailed information is available in the release notes for
+[Trino 387](https://trino.io/docs/current/release/release-387.html),
+[Trino 388](https://trino.io/docs/current/release/release-388.html),
+[Trino 389](https://trino.io/docs/current/release/release-389.html),
+[Trino 390](https://trino.io/docs/current/release/release-390.html),
+and [Trino 391](https://trino.io/docs/current/release/release-391.html)
+
+## Concept of the episode: Polymorphic table functions
+
+We normally cover a broad variety of topics in the Trino community broadcast,
+exploring different technical details, pull requests, and neat things that are
+going on in Trino at large. This episode, however, we're going to be more
+focused, only taking a look at a particular piece of functionality that we're
+all very excited about: polymorphic table functions, or PTFs for short. If
+you're unfamiliar with what this means, that can sound like technobabble word
+soup, so we can start exploring this with a simple question...
+
+### What is a table function?
+
+The easiest answer to this question is that it's a function which returns a
+table. Scalar, aggregate, and window functions all work a little differently,
+but ultimately, they all return a single value each time they are invoked. Table
+functions are unique in that they return an entire table. This gives them some
+interesting properties that we'll dive into, but it also means that you can only
+invoke them in situations where you'd use a full table, such as a `FROM` clause:
+
+```sql
+SELECT
+  *
+FROM
+  TABLE(my_table_function('foo'));
+```
+
+You can also use table functions in joins:
+
+```sql
+SELECT
+  *
+FROM
+  TABLE(my_table_function('bar'))
+JOIN
+  TABLE(another_table_function(1, 2, 3))
+ON true;
+```
+
+And while that's all neat, it begs the question...
+
+#### What can you do with table functions?
+
+So... even standard table functions are cool. They have to return a pre-defined
+schema, which limits their flexibility, but they still have some interesting
+uses as means of shortening queries or performing multiple operations at once.
+If you frequently find yourself selecting from the same table with a `WHERE`
+clause checking equality to a specific column but with a different value each
+time, you could define a table function which takes that value as a parameter
+and allows you to skip all the copying and pasting just for the sake of one line
+changing. You could take an extremely lengthy sub-query with multiple joins and
```

Comment:
I'm not sure that this is the kind of usage of PTF we want to suggest.
Such a function would likely be far less efficient that a query with multiple joins executed directly in Trino. And it might be a big challenge to write. In general, if something can be expressed in SQL, it is better to write a normal query, and let Trino optimize and execute it. (However, if someone have got a very specific use case, which they want to optimize in a specific way, writing a PTF might be a good idea)

We're planning to optimize the PTF operator when we have it, to make it a competitive alternative to other operators, but it won't happen soon.

Reply:
I'd like to keep the example as a really basic one - do you think it's reasonable to leave the example in as a sample of what _can_ be done, but not what _should_ be done?

Reply:
I also didn't realize when commenting this morning - this is the section describing an ordinary (non-polymorphic) table function, which I imagine is less of a concern in terms of performance?

Reply:
>I also didn't realize when commenting this morning - this is the section describing an ordinary (non-polymorphic) table function, which I imagine is less of a concern in terms of performance?

Polymorphic functions are more expressive than monomorphic, but both types are executed in the same way, with the same performance challenges. Also, from the user's perspective, they are invoked in the same way.